### PR TITLE
Use the same network for allocation as for the physical layer

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/1388.

This also fixes a bug I found: In allocation the `max_flow_rate` of a node connecting the main network to a subnetwork was not taken into account

### CODE REVIEW

File: core/src/allocation_init.jl [hunk 1]
```diff
@@ -460,27 +243,29 @@ Only finite capacities get a constraint.
...
         if !isinf(c) && edge ∉ main_network_source_edges
             push!(edge_ids_finite_capacity, edge)
         end
     end
+
     problem[:capacity] = JuMP.@constraint(
         problem,
         [edge = edge_ids_finite_capacity],
-        F[edge] <= capacity[edge...],
+        F[edge] ≤ capacity[edge...],
```
Comment (visr) [inline core/src/allocation_init.jl:268]: Do you use a font with coding ligature support? To me a `<` followed by a `=` already looks like one symbol. I use Cascadia Code, see https://github.com/microsoft/cascadia-code?tab=readme-ov-file#font-features.

If you use that perhaps we can just keep the ASCII version?

File: core/src/allocation_init.jl [hunk 2]
```diff
@@ -666,37 +341,88 @@ The constraint indices are Basin node IDs.
...
+
+    edges_allocation = only(F.axes)
+
+    for node_id in node_ids
+
+        # No flow conservation constraint on sources/sinks
+        is_source_sink = node_id.type in
+        [NodeType.FlowBoundary, NodeType.LevelBoundary, NodeType.UserDemand]
+
+        # No flow conservation on nodes with fractional flow outneighbors
```
Comment (visr) [inline core/src/allocation_init.jl:368]:
```suggestion
        # No flow conservation on nodes with FractionalFlow outneighbors
```

File: core/src/allocation_init.jl [hunk 3]
```diff
@@ -666,37 +341,88 @@ The constraint indices are Basin node IDs.
...
+            if edge_in in edges_allocation
+                push!(inflows_node, F[edge_in])
+            end
+            edge_out = (node_id, neighbor_id)
+            if edge_out in edges_allocation
+                push!(outflows_node, F[edge_out])
+            end
+        end
+
+        # If the node is a basin, add basin in- and outflow
```
Comment (visr) [inline core/src/allocation_init.jl:395]:
```suggestion
        # If the node is a Basin, add Basin in- and outflow
```

**Review by visr — APPROVED** [review]
Such a nice simplification!

File: core/src/allocation_optim.jl [hunk 4]
```diff
@@ -829,63 +827,62 @@ function save_allocation_flows!(
...
+            edge_metadata = graph[edge_1...]
+        else
+            edge_1_reverse = reverse(edge_1)
+            flow_rate -= JuMP.value(F[edge_1_reverse])
+            sign_2 = 1.0
+            edge_metadata = graph[edge_1_reverse...]
+        end
+
+        # Check whether the next edge is the current one reversed
+        if edge_2 == reverse(edge_1)
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/allocation_optim.jl:863]: That's an excellent question. I actually caught a bug looking into this. If an edge is bidirectional, both allocation edges are created consecutively:

https://github.com/Deltares/Ribasim/blob/dccb31eec6ff4592fa3bfd54eaf48ff98e5deee8/core/src/allocation_init.jl#L64-L73

Also, the underlying structure of `capacity` as a `SparseAxisArray` is an `OrderedDict`, so I think this is safe?

**Pull request merged**